lessons/cryptography/cryptosystem-types/cryptosystems-types.md

```diff
@@ -0,0 +1,90 @@
+
+### RSA
+Rivest-Shamir-Adleman (RSA) is a type of asymmetric encryption that uses a private and public key based on two large prime numbers. The algorithm for encryption involves multiplying two very large prime numbers, and is considered very hard to crack because recovering the two original prime numbers is a mathematical problem that is currently thought to be infeasible to solve efficiently. There are some known attacks against RSA, but many of these attacks occur when small prime numbers are used. For example, if small prime numbers are used, an attacker could create a dictionary by encrypting high probability plaintext messages using various keys, and compare the resulting ciphertext with a stolen ciphertext message. If the messages match, the attacker now knows the original prime numbers used to encrypt the message. 
+
+### Diffie-Hellman
+The Diffie-Hellman key exchange process involves the use of asymmetric encryption, and is used to establish a shared private key for use in future communications. When we previously talked about AES (symmetric cryptosystem), we noted that Alice and Bob would need a way to securely share a private key. The Diffie-Hellman key exchange process can accomplish this. The Diffie-Hellman algorithm involves a public modulus and base value, and two secret values chosen by Alice and Bob. Through an elegant mathematical algorithm, Alice and Bob can each use their own private secret values to calculate a public key that can be sent to each other unencrypted. They can then each use the other person's public key and their own secret value to calculate the same secret key. 
+
+The strength of the Diffie-Hellman key exchange process depends on using a strong random number generator to choose the public modulus so that it cannot be easily predicted by an attacker. In addition, this process does not provide any method of authentication between Alice and Bob, so it is vulnerable to a man-in-the-middle attack where an attacker could intercept information between Alice and Bob and instead establish two separate Diffie-Hellman key exchanges with Alice and Bob. 
+
+### Elliptic Curve Cryptosystem
+Elliptic Curve Cryptosystems (ECC) are the most powerful type of cryptography currently used today. ECC is based on a mathematical concept called an elliptic curve, which is a set of points that satisfy a specific mathematical equation. For example, an elliptic curve is shown below. 
+
+![Sample Elliptic Curve (https://blog.cloudflare.com/a-relatively-easy-to-understand-primer-on-elliptic-curve-cryptography/)](.rsrc/elliptic-curve.png)
+
+ECC is based on solving a challenging mathematical problem that is considered more difficult to solve and more computationally expensive than the mathematical problem that RSA and Diffie-Hellman are based on. With ECC, smaller keys will achieve the same level of security compared to RSA. 
```

Comment:
Maybe mention that this also makes it more efficient for the same level of security. Also maybe mention some implementations such as ECDH rather than just standard DH etc.

Reply:
Done